### PR TITLE
*: support certificate based authentication

### LINKED ISSUE #9708: Certificate based authentication
## Feature Request

**Is your feature request related to a problem? Please describe:**

MySQL supports certificate-based authentication.  This currently not supported by TiDB.

**Describe the feature you'd like:**

Ability to auth without passwords.

**Describe alternatives you've considered:**

N/A

**Teachability, Documentation, Adoption, Migration Strategy:**

Same as in MySQL.

### PR DESCRIPTION
**Pull request by lysu**
### What problem does this PR solve? 

fixes #9708

to make TiDB  support certificate-based authentication like mysql.

### What is changed and how it works?

- add new `mysql.global_priv` system table
- maintain `require clause` info in those new table
- retrieve cert info from `tlsConn` 
- check current user's `ssl_type`, `ssl_issuer`, `ssl_subject` - (WIP)

### Check List 

Tests 

 - Unit test
 - Integration test(WIP)

Code changes

 - impl change

Side effects

 - N/A

Related changes

 - Need to cherry-pick to the release branch - 3.0

Release note

 - support certificate based authentication

---
This change is [Reviewable](https://reviewable.io/reviews/pingcap/tidb/13955)

### CODE REVIEW

File: privilege/privileges/cache.go [hunk 1]
```diff
@@ -584,12 +585,15 @@ func (p *MySQLPrivilege) decodeGlobalPrivTableRow(row chunk.Row, fs []*ast.Resul
 				var privValue GlobalPrivValue
 				err := json.Unmarshal(hack.Slice(privData), &privValue)
 				if err != nil {
-					return errors.Trace(err) // TODO: ingore it ?
+					logutil.BgLogger().Error("one userglobal priv data is broken, forbidden login until data be fixed",
```
Comment (tiancaiamao) [inline privilege/privileges/cache.go]: s/userglobal/user global

File: executor/grant.go [hunk 2]
```diff
@@ -134,6 +147,20 @@ func (e *GrantExec) Next(ctx context.Context, req *chunk.Chunk) error {
 	return nil
 }
 
+// checkAndInitGlobalPriv checks if global scope privilege entry exists in mysql.global_priv.
+// If unexists, insert a new one.
```
Comment (jackysp) [inline executor/grant.go]:
```suggestion
// If not exist, insert a new one.
```

File: server/tidb_test.go [hunk 3]
```diff
@@ -22,6 +22,7 @@ import (
 	"crypto/x509"
 	"crypto/x509/pkix"
 	"encoding/pem"
+	"github.com/pingcap/tidb/util"
```
Comment (jackysp) [inline server/tidb_test.go]: fix fmt

**Comment by lysu** [comment]
friendly ping @imtbkcat @jackysp @tiancaiamao

File: executor/grant.go [hunk 4]
```diff
@@ -211,25 +245,93 @@ func initColumnPrivEntry(ctx sessionctx.Context, user string, host string, db st
...
+
+var emptyGP = privileges.GlobalPrivValue{SSLType: privileges.SslTypeNotSpecified}
+
+func tlsOption2GlobalPriv(tlsOptions []*ast.TLSOption) (priv []byte, err error) {
+	if len(tlsOptions) == 0 {
+		priv = []byte("{}")
+		return
+	}
+	gp := privileges.GlobalPrivValue{SSLType: privileges.SslTypeNotSpecified}
+	for _, tlsOpt := range tlsOptions {
```
Comment (imtbkcat) [inline executor/grant.go]: Duplicate declare should not be allowed here, like;
`grant all on *.* to tester require cipher 'TLS_AES_256_GCM_SHA384' cipher 'RC4-SHA';`

File: executor/grant.go [hunk 5]
```diff
@@ -86,9 +90,14 @@ func (e *GrantExec) Next(ctx context.Context, req *chunk.Chunk) error {
...
 
 		// If there is no privilege entry in corresponding table, insert a new one.
-		// DB scope:		mysql.DB
-		// Table scope:		mysql.Tables_priv
-		// Column scope:	mysql.Columns_priv
+		// Global scope:		mysql.global_priv
+		// DB scope:			mysql.DB
+		// Table scope:			mysql.Tables_priv
+		// Column scope:		mysql.Columns_priv
+		err = checkAndInitGlobalPriv(e.ctx, user.User.Username, user.User.Hostname)
```
Comment (imtbkcat) [inline executor/grant.go]: just check at `e.TLSOptions != nil` ?

**Comment by lysu** [comment]
@tiancaiamao @jackysp  need another LGTM, PTAL :smile:

**Pull request merged**